### DISCUSSION TITLE
How to validate multiple types in pydantic

### DISCUSSION THREAD

**Discussion by 0num4** [pydantic/pydantic#9090] [Question] [answered]
How can I convert a dictionary to a Pydantic class using something like TypeAdapter when there are multiple pydantic.BaseModel classes, and there are classes AINT and ASTR with the same property a being int and str, respectively?

**Comment by 0num4** [comment]
I've created a function like the one below, but I can't determine the correct type using isinstance, and since the hand_cards attribute exists in other instances as well, it gets caught by the first condition.
```python
from typing import Literal, Union, TypeAlias
import pydantic

HandEventRecordDataLiteral: TypeAlias = Union[
    PaifuTypeInternalPydantic.HandInfo,  # hand_cards
    PaifuTypeInternalPydantic.OperationInfo,  # in_card
    PaifuTypeInternalPydantic.ActionInfo,  # action
    PaifuTypeInternalPydantic.OutCardInfo,  # out_card
]


def assignDictToHandEventRecordDataLiteral(data: dict) -> HandEventRecordDataLiteral:
    if "hand_cards" in data:
        return PaifuTypeInternalPydantic.HandInfo(**data)
    elif "in_card" in data:
        return PaifuTypeInternalPydantic.OperationInfo(**data)
    elif "action" in data:
        return PaifuTypeInternalPydantic.ActionInfo(**data)
    elif "out_card" in data:
        return PaifuTypeInternalPydantic.OutCardInfo(**data)
```

**Comment by samuelcolvin** [comment]
I don't understand the question, please give an example.

**Reply by 0num4** [reply]
thanks for reply.
I add example code.

**Reply by samuelcolvin** [reply, marked as answer]
You should use a functional discriminator- https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-callable-discriminator.
